system/security/plugins/testauthSecurity/testauthSecurity.cpp

```diff
@@ -0,0 +1,591 @@
+            {
+                access = (SecAccessFlags) newAccess;
+                first = false;
+            }
+        }
+        return getFileScopeAccess(element, nodes, ++pos, access, first);
+    }
+
+    SecAccessFlags queryECLWUScopeAccess(const char* resource)
+    {
+        int* allow = eclwuScopeAccesses.getValue(resource);
+        return allow ? (SecAccessFlags) *allow : defaulteECLWUScopeAccess;
+    }
+};
+
+typedef CUserAccess* CUserAccessPtr;
+
+class CTestAuthSecurityManager : public CBaseSecurityManager
+{
+    static const SecFeatureSet s_implementedFeatureMask = SMF_QuerySecMgrType | SMF_QuerySecMgrTypeName | SMF_CreateAuthMap |
+                                                          SMF_CreateFeatureMap | SMF_LogoutUser | SMF_GetDescription | SMF_Authorize |
+                                                          SMF_AuthorizeEx_Named | SMF_GetAccessFlagsEx |
+                                                          SMF_AuthorizeFileScope_Named | SMF_AuthorizeWorkUnitScope_Named;
+    static const SecFeatureSet s_safeFeatureMask = s_implementedFeatureMask | SMF_CreateSettingMap;
+
+    MapStringTo<CUserAccess*> userAccessMap;
+
+    void readUserAccess(IPropertyTree* secMgrCfg, IPropertyTree* bindConfig)
+    {
+        StringBuffer s;
+        Owned<IPropertyTreeIterator> iter = secMgrCfg->getElements("useraccess");
+        ForEach(*iter)
+        {
+            IPropertyTree& userSettings = iter->query();
+
+            StringBuffer userName, password;
+            if (!readUserNamePassword(userSettings, userName, password))
+                continue;
+
+            CUserAccessPtr* userAccess = userAccessMap.getValue(userName);
+            if (userAccess)
+                continue;
+
+            Owned<CUserAccess> newUserAccess = new CUserAccess(userName, password);
+            readDefaultAccess(userSettings, newUserAccess);
+            readFeatureAccess(bindConfig->queryProp("@serviceType"), userSettings, newUserAccess);
+            readFileScopeAccess(userSettings, newUserAccess);
+            readECLWUScopeAccess(userSettings, newUserAccess);
+            userAccessMap.setValue(userName.str(), newUserAccess.getClear());
+        }
+    }
+
+    bool readUserNamePassword(IPropertyTree& userSettings, StringBuffer& userName, StringBuffer& password)
+    {
+        const char* secretKey = userSettings.queryProp("@secretkey");
+        if (!isEmptyString(secretKey))
+        {
+            const char* vaultKey = userSettings.queryProp("@vaultkey");
+            Owned<IPropertyTree> secretTree;
+            if (isEmptyString(vaultKey))
+                secretTree.setown(getSecret("authn", secretKey));
+            else
+                secretTree.setown(getVaultSecret("authn", vaultKey, secretKey, nullptr));
+            if (!secretTree)
+                throw makeStringExceptionV(-1, "Error retrieving the secret for %s.", secretKey);
+
+            getSecretKeyValue(userName, secretTree, "username");
```

Comment:
question: @afishbeck would you expect the username to be in the secret?  I was expecting only the password.